README.md

```diff
@@ -0,0 +1,178 @@
+
+Depending on that method arguments and return parameters the hooks and the
+wrapped method receive different values. See the inlined comments for notable
+differences.
+
+```go
+func (mw *InterfaceMiddleware) ProcessOne(arg0 string) (string, error) {
+	methodInfo := types.MethodInfo{
+		Package:   "<full-package-path>",
+		Interface: "Interface",
+		Method:    "ProcessTwo",
+	}
+
+	// ProcessOne doesn't accept context so one is created
+	ctx := context.Background()
+
+	ctx = mw.hook.OnEntry(ctx, methodInfo)
+	res0, res1 := mw.next.ProcessOne(arg0)
+	// ProcessOne returns an error as last return parm and it's passed to the OnExit hook
+	mw.hook.OnExit(ctx, methodInfo, res1)
+	return res0, res1
+}
+
+func (mw *TwoMethodsWithArgAndReturnMiddleware) ProcessTwo(arg0 context.Context, arg1 int) bool {
+	methodInfo := types.MethodInfo{
+		Package:   "<full-package-path>",
+		Interface: "Interface",
+		Method:    "ProcessTwo",
+	}
+
+	// ProcessTwo accepts a context and it is used
+	ctx := arg0
+
+	ctx = mw.hook.OnEntry(ctx, methodInfo)
+	res0 := mw.next.ProcessTwo(ctx, arg1)
+	// ProcessTwo doesn't return an error type so OnExit always receives nil
+	mw.hook.OnExit(ctx, methodInfo, nil)
+	return res0
+}
+
+```
+
+All hooks accept a `Context` they can extract values from as well as `MethodInfo`, containing
+information about the method being invoked. Some hooks accept additional parameters also
+return a value. These are covered below.
+
+If the Interface method accepts a `context.Context` as its first parameter it will be passed
+to all hooks and the wrapped method. If the first parameter is not a context, a
+`context.Background()` context will be instantiated and passed to all hooks instead.
+
+### OnEntry
+
+```go
+OnEntry(context.Context, MethodInfo) context.Context
+```
+
+This hook is called just before the wrapped method is called.
+
+It returns a `context` that is passed to the rest of the hooks as well as to the
+wrapped method. Hook implementations can use this to store and communicate values
+within a method call context (e.g. span start/end, processing duration etc.)
+
+### OnExit
+
+```go
+mw.hook.OnExit(context.Context, MethodInfo, error)
+```
+
+This hook is called right after the wrapped method returns. If the wrapped
+method's last return parameter was of type `error` it is passed to the OnExit
+method. If not - OnExit gets nil.
+
+# Usage
+
+## Generating the Middleware
+
+### Command line
+
+```sh
+go run github.com/Makpoc/gomigen/cmd/gomigen . InterfaceName
+```
+
+The destination directory where the generated package with the middleware will be saved
+can be provided as an optional parameter.
+
+```sh
+go run github.com/Makpoc/gomigen/cmd/gomigen -o ../mw . InterfaceName
+```
+
+The generated file will import this module's `types` packages so make sure to add that
+package to the `tools.go` file (see below).
+
+### Go Generate
+
+Simply inject a go comment like you would with counterfeiter or gentools
```

Comment:
NIT: Maybe add links to `counterfeiter` and `gentools`.

Reply:
I've reworded this sentence in the readme, but I did mention both projects as inspiration and added links earlier in the file.